### ISSUE TITLE
Old comments from deleted test case runs appearing in newly created test case runs

### ISSUE THREAD

**Issue by cloudbloat** [kiwitcms/Kiwi#1028]
I used to have thousands of test case runs with many comments in each... I created simple selenium script, which deleted these test case runs through gui (Login to Kiwi, filter test case runs, click delete, confirm it. Again.).
Deletion was ok but now when I create new test case run, right after it is created with old comments from deleted runs in it....

I am unable to get into database to clear comments, and not sure if it would be enough..
![Kiwin_comments](https://user-images.githubusercontent.com/53474302/62123548-735a1e80-b2c8-11e9-84e7-c357069563c6.jpg)

**Comment by cloudbloat** [comment]
I am now deleting all comments for all test cases using fabricated post requests to https://kiwi.sysct.cz/comments/delete/   .... and will let you know, it it works correctly afterwards...

**Comment by atodorov** [comment]
> I am now deleting all comments for all test cases using fabricated post requests to
> https://kiwi.sysct.cz/comments/delete/ .... and will let you know, it it works correctly afterwards...

@cloudbloat any update here?

**Comment by cloudbloat** [comment]
> I am now deleting all comments for all test cases using fabricated post requests to
> https://kiwi.sysct.cz/comments/delete/ .... and will let you know, it it works correctly afterwards...
> 
> @cloudbloat any update here?

Thanks for interest.

 After deleting all comments manually it works perfectly now, IDs are now not overlapping and each new comment has new ID (it doesn't recycle IDs, but increase them). I dont know what caused it, maybe incorrect usage from my side.

**Comment by cloudbloat** [comment]
So it happened again after restarting our kiwi server...I will try to investigate more and provide more informations....

**Comment by cloudbloat** [comment]
@atodorov may I ask if there is way to see all existing comments in KIWI mariaDB ? To better investigate problem...(I searched but couldnt find a way)

**Comment by atodorov** [comment]
```
$ ./manage.py shell
>>> from django_comments.models import Comment

>>> Comment.objects.all()

>>> for comment in Comment.objects.all():
...     print(comment.submit_date, comment.comment)
... 
```

**Comment by cloudbloat** [comment]
So I have deleted all test runs - running and stopped ones, and thanks to your help, I am able to find that there still many comments exists....

Shouldnt they all be deleted with the test runs ? 

(I insert the comments using: rpc_client.exec.TestExecution.add_comment(int(test_case_run_id), str(comment)))

**Comment by atodorov** [comment]
Can you post the script which creates these comments? Something isn't right here but without a reproducer there's nothing we can do to debug and figure it out.

**Comment by cloudbloat** [comment]
Thanks for interest... I am inserting comment using Python: 

**rpc_client.exec.TestExecution.add_comment(int(test_case_run_id), str(comment))**

It didnt happen few days after restarting Kiwi instance, but today (after about 10 days of usage) it appeared again...

**Comment by atodorov** [comment]
Several notes:

1) tcms_api-5.3 is rather old. There are new versions but you should take into account that newer versions are compatible with newer backend versions only. This is all documented but shouldn't matter in your case.

2) `TestExecution.add_comment` is apparently working since you can see the comments

3) Can you confirm that when your scripts are running they are using `test_case_run_id==1038` (later renamed to execution_id) ? In other words, when the above code snippet executes do you see the newly added comment in the web UI?

4) Running scripts(test jobs) in parallel shouldn't matter in this case. True you may have a race condition where the TR or TE status is overwritten but this wouldn't prevent multiple comments from being added. They are just saved into another DB table and later queried from there, so multiple comments would be shown if there have been multiple executions.

5) The fact that you say you are restarting your Kiwi TCMS instance doesn't matter. These things are in the database. Either they are there or they are not.

We need more information here:

0) Version of Kiwi TCMS you are running, DB engine and version.

1) with browser tools inspect the delete comment icon. In the HTML you will notice hidden input fields for `object_pk` and `comment_id`. Make a note of them. `object_pk` should be the same as `test_case_run_id`. `comment_id` is the ID of this record in the comments table.

When you see comments, which you think must not be there inspect both of the IDs for them and post the findings here.

2) Your original report says you've first created test cases and then deleted them via the UI. Here comments relationship is not a FK (this is how django-comments works) so comments are not deleted via cascade delete.  

The only thing I can imagine that may be happening is that originally you've created many test runs/executions and comments and now after the deletion your TE (TestCaseRun in older versions) objects are being created with the same IDs and these older comments get assigned to them.

I can't reproduce this behavior, i.e. when I delete something the next object is created with an ID which is +1 of the old ID.


Notice: I do see something strage in my exploration. `object_pk` mattches the `case_id` instead of the execution id. This is what is shown in the HTML, not sure ATM if it has always been like this or we've broken it somehow. This is why I need you to provide more info.

**Comment by cloudbloat** [comment]
1.	tcms_api-5.3 is rather old. There are new versions but you should take into account that newer versions are compatible with newer backend versions only. This is all documented but shouldn't matter in your case.
> OK

2.	TestExecution.add_comment is apparently working since you can see the comments
> YES WORKS FINE, comments are added in 95% cases.

3.	Can you confirm that when your scripts are running they are using test_case_run_id==1038 (later renamed to execution_id) ? In other words, when the above code snippet executes do you see the newly added comment in the web UI?
> YES WORKS OK. IT REALLY LOOKS LIKE IT WORKS UNTIL STARTED IN PARALLEL, THEN SOME COMMENTS ARE BEING LOST. - I will be reproducing today (26.9.) to find out exact state and add debug messages on "add comment" function.

4.	Running scripts(test jobs) in parallel shouldn't matter in this case. True you may have a race condition where the TR or TE status is overwritten but this wouldn't prevent multiple comments from being added. They are just saved into another DB table and later queried from there, so multiple comments would be shown if there have been multiple executions.
> Paralel execution from one test case run is used only when debugging the test scenario...It is not standard way of usage.

5.	The fact that you say you are restarting your Kiwi TCMS instance doesn't matter. These things are in the database. Either they are there or they are not.
OK :)



Q/A:
1.	Version of Kiwi TCMS you are running, DB engine and version.
- Kiwi TCMS - 6.10
- DB engine + version - Server version: 5.5.60-MariaDB MariaDB Server;  
- "SELECT PLUGIN_NAME, PLUGIN_VERSION, PLUGIN_TYPE_VERSION, PLUGIN_LIBRARY     , PLUGIN_LIBRARY_VERSION, PLUGIN_AUTHOR FROM information_schema.PLUGINS WHERE PLUGIN_NAME = 'innodb';" >> InnoDB - PLUGIN_VERSION: 5.5, PLUGIN_TYPE_VERSION: 50560.0
- "SHOW GLOBAL VARIABLES LIKE 'innodb_ver%';" >> innodb_version | 5.5.59-MariaDB-38.11 
- mariadb-server-5.5.60-1.el7_5.x86_64
- mariadb-libs-5.5.60-1.el7_5.x86_64
- mariadb-5.5.60-1.el7_5.x86_64

2.	with browser tools inspect the delete comment icon. In the HTML you will notice hidden input fields for object_pk and comment_id. Make a note of them. object_pk should be the same as test_case_run_id. comment_id is the ID of this record in the comments table.
- please see following image (object_pk differs from test run ID):
![kiwi_comments_detail](https://user-images.githubusercontent.com/53474302/65671619-d7d6f700-e047-11e9-88b1-4988394eaa0a.jpg)
+
![kiwi_more_details](https://user-images.githubusercontent.com/53474302/65672211-effb4600-e048-11e9-9f2b-f70d2a61f395.png)

(> The only thing I can imagine that may be happening is that originally you've created many test runs/executions and comments and now after the deletion your TE (TestCaseRun in older versions) objects are being created with the same IDs and these older comments get assigned to them.
- **ABSOLUTELY thats it.**)

**Comment by cloudbloat** [comment]
So now I am able to always reproduce, but cannot find exact steps to reproduce, my STR:
1. Start test case run in more processes (so each process is sole script which autenthicates and connects to xmlrpc, send comments) >> not all of them are visible in the GUI.

**Comment by atodorov** [comment]
There are a number of issues here: 

1) The wrong object_pk hidden field value in HTML templates. This will actually be removed b/c we don't really use it when deleting comments. Only comment_id is used.

2) Comments not actually deleted but marked as hidden. We'll now call a .delete() on the query set and actually remove these objects from the DB

3) Comments not removed when the object they point to is removed. This will be resolved via pre-delete signal b/c we already have this machinery in place and b/c we can't reliably modify django_comments to turn the object_pk field into a ForeignKey relationship (b/c it points to multiple other tables). Again this will be a permanent delete from the DB.

1-3 will be solved by https://github.com/kiwitcms/Kiwi/pull/1220 which will also close this issue.


4) As stated earlier you may have had a situation where PK values have been recycled. That is an existing object has been deleted, then a new one created with the same ID which manifested as old comments being shown. 

We still don't have enough information to figure out why this was happening in the first place. The only pointers I can give you is to check how transactions on your MariaDB have been configured. Maybe also look at sequences and auto-increment (LAST_INSERT_ID) settings. I can't reproduce and it's outside the scope of this issue anyway. 

With the merging of deletion changes even if IDs are recycled the problem of seeing older comments should be less prominent. Will be non existent if you manually remove all comments for which object_pk doesn't point anywhere.

5) Possible issues with parallel script execution : I don't see issues with adding comments in parallel per se. Order of comments & TE statuses may be misaligned due to timing (race conditions) but this is the nature of manipulating objects in parallel. Adding comments is an insert operation so nothing should be lost. If RPC didn't report an exception then there wasn't any when performing the database operation.

In your last comment you also state

> play with it to reproduce - like ... deleting it while all tests are not finished

If you have operations pending towards an existing object and you delete that same object from the DB then the behavior is undefined. Transactional settings on your DB engine will likely have effect here. This may or may not be a contributing factor to issue 4). In any case I don't think there's something we can do about it on our end. 


If you continue to see 4) and 5) please collect more debugging information & steps to reproduce and open another issue.

**Comment by cloudbloat** [comment]
Fix Confirmed - Thank you very much !!! Now it works without described problems :)
